### PR TITLE
Expose indentation suppressor from SmartIndenter

### PR DESCRIPTION
**Pull request by SaschaNaz**
Separated from #4609.

`shouldIndentChildNode` function already have indentation suppressing feature, but currently auto-formatter disables it to get delta value and implements another suppressor in `computeIndentation` function.

This PR split the suppressor from `shouldIndentChildNode` and expose it as a separate function so that formatter can use it without implementing its own one.

The function will help suppress indentation of:
* `ImportClause` node in `ImportDeclaration` node. (#4609)
* `JsxClosingElement` node. (#4398)
* false branch in ternary operator. (#5905 which already includes this PR)

Note: `shouldInheritParentIndentation` and its internal `nodeWillIndentChild` receives TextRangeWithKind to check more conditions other than syntax kind.

### CODE REVIEW

File: src/services/formatting/formatting.ts [hunk 1]
```diff
@@ -282,19 +282,19 @@ namespace ts.formatting {
      */
     function getOwnOrInheritedDelta(n: Node, options: FormatCodeOptions, sourceFile: SourceFile): number {
         let previousLine = Constants.Unknown;
-        let childKind = SyntaxKind.Unknown;
+        let child: Node = null;
```
Comment (DanielRosenwasser) [inline src/services/formatting/formatting.ts]: Use `undefined`, not `null`

Comment (SaschaNaz) [inline src/services/formatting/formatting.ts]: Fixed.

File: src/services/formatting/formatting.ts [hunk 2]
```diff
@@ -503,7 +490,7 @@ namespace ts.formatting {
                             indentation -= options.IndentSize;
                         }
 
-                        if (SmartIndenter.shouldIndentChildNode(node.kind, SyntaxKind.Unknown)) {
+                        if (SmartIndenter.shouldIndentChildNode(node, null)) {
```
Comment (DanielRosenwasser) [inline src/services/formatting/formatting.ts]: `undefined` not `null`

File: src/services/formatting/formatting.ts [hunk 3]
```diff
@@ -493,9 +480,9 @@ namespace ts.formatting {
                     }
                 },
                 getIndentation: () => indentation,
-                getDelta: () => delta,
+                getDelta: (child) => (!delta || SmartIndenter.shouldInheritParentIndentation(node, child)) ? 0 : delta,
```
Comment (DanielRosenwasser) [inline src/services/formatting/formatting.ts]: I'm a little confused about the condition here - I would think that you want to maintain the delta if you're inheriting from the parent.

Comment (SaschaNaz) [inline src/services/formatting/formatting.ts]: The delta value increases indentation for child node, so it should be zero when the child node should just copy indentation from parent.

**Comment by RyanCavanaugh** [comment]
Ping @vladima

File: src/services/formatting/formatting.ts [hunk 4]
```diff
@@ -493,9 +480,9 @@ namespace ts.formatting {
                     }
                 },
                 getIndentation: () => indentation,
-                getDelta: () => delta,
+                getDelta: (child) => (delta && SmartIndenter.shouldInheritParentIndentation(node, child)) ? 0 : delta,
```
Comment (vladima) [inline src/services/formatting/formatting.ts]: to me condition ` SI.shouldInheritParentIndentation() ? 0 : delta` look more clear and it is doing the same thing.
- if delta is not-null-or-zero then body of `getDelta` boils down to `SI.shouldInheritParentIndentation() ? 0 : delta`
- if delta zero then `getDelta` evaluates to just `delta` which will mean the same

Comment (SaschaNaz) [inline src/services/formatting/formatting.ts]: You're right, it's doing the same thing. I was to pass the SI check when delta value is already zero but I agree that it looks less clear. Do you want me to fix this?

File: src/services/formatting/formatting.ts [hunk 5]
```diff
@@ -511,6 +498,13 @@ namespace ts.formatting {
                         }
                     }
                 },
+                getCopyForSpecificChild(child) {
+                    if (!delta) {
+                        // delta value is already 0, so do not copy
+                        return this;
```
Comment (vladima) [inline src/services/formatting/formatting.ts]: we rarely use manipulations that involve `this` in compiler/language service. I'd prefer if we can lift members of object literal to functions and then just call functions

File: src/services/formatting/formatting.ts (hunk 4 above)
Comment (vladima) [inline src/services/formatting/formatting.ts]: yes, please

Comment (SaschaNaz) [inline src/services/formatting/formatting.ts]: Done!

File: src/services/formatting/formatting.ts (hunk 5 above)
Comment (SaschaNaz) [inline src/services/formatting/formatting.ts]: Okay, now there is no `this` :D

File: src/services/formatting/formatting.ts [hunk 6]
```diff
@@ -489,13 +471,13 @@ namespace ts.formatting {
                             return indentation;
                         default:
                             // if token line equals to the line of containing node (this is a first token in the node) - use node indentation
-                            return nodeStartLine !== line ? indentation + delta : indentation;
+                            return nodeStartLine !== line ? indentation + getEffectiveDelta(delta, container) : indentation;
                     }
                 },
                 getIndentation: () => indentation,
-                getDelta: () => delta,
+                getDelta: (child) => getEffectiveDelta(delta, child),
```
Comment (DanielRosenwasser) [inline src/services/formatting/formatting.ts]: remove parens around `child`

Comment (SaschaNaz) [inline src/services/formatting/formatting.ts]: Done!

File: src/services/formatting/smartIndenter.ts [hunk 7]
```diff
@@ -465,12 +465,10 @@ namespace ts.formatting {
...
             return false;
         }
-
-        export function shouldIndentChildNode(parent: SyntaxKind, child: SyntaxKind): boolean {
-            if (nodeContentIsAlwaysIndented(parent)) {
-                return true;
-            }
-            switch (parent) {
+        
+        export function nodeWillIndentChild(parent: TextRangeWithKind, child: TextRangeWithKind, indentByDefault: boolean) {
```
Comment (mhegazy) [inline src/services/formatting/smartIndenter.ts]: this should be marked as /*@internal*/

**Comment by mhegazy** [comment]
sorry for the delay. Can you add the @internal annotation in your next PR.

**Pull request merged**